File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.java
class Solution {
    // Parallel arrays to completely eliminate Node objects and flatten the memory footprint
    private char[] leftChar;
    private char[] rightChar;
    private int[] prefix;
    private int[] suffix;
    private int[] max;
    private int[] length;

    public int[] longestRepeating(String s, String queryCharacters, int[] queryIndices) {
        int n = s.length();
        int k = queryIndices.length;

        // Using exactly 4 * n elements for flat, cache-friendly primitive segment tree representation
        int treeSize = 4 * n;
        leftChar = new char[treeSize];
        rightChar = new char[treeSize];
        prefix = new int[treeSize];
        suffix = new int[treeSize];
        max = new int[treeSize];
        length = new int[treeSize];

        // Avoid repeated string extraction wrappers in inner hot loops
        char[] sArr = s.toCharArray();
        char[] qChars = queryCharacters.toCharArray();

        build(1, 0, n - 1, sArr);

        int[] answer = new int[k];
        for (int i = 0; i < k; i++) {
            update(1, 0, n - 1, queryIndices[i], qChars[i]);
            answer[i] = max[1]; // The global maximum answer is always stored at root node index 1
        }

        return answer;
    }

    private void build(int node, int left, int right, char[] s) {
        if (left == right) {
            initLeaf(node, s[left]);
            return;
        }

        int mid = left + (right - left) / 2;
        int leftChild = node << 1;       // Optimized bitwise left-shift (node * 2)
        int rightChild = leftChild | 1;   // Optimized bitwise bit-or (node * 2 + 1)

        build(leftChild, left, mid, s);
        build(rightChild, mid + 1, right, s);

        merge(node, leftChild, rightChild);
    }

    private void update(int node, int left, int right, int index, char ch) {
        if (left == right) {
            initLeaf(node, ch);
            return;
        }

        int mid = left + (right - left) / 2;
        int leftChild = node << 1;
        int rightChild = leftChild | 1;

        if (index <= mid) {
            update(leftChild, left, mid, index, ch);
        } else {
            update(rightChild, mid + 1, right, index, ch);
        }

        merge(node, leftChild, rightChild);
    }

    // Inline leaf node initialization to prevent any object/variable allocation footprint
    private void initLeaf(int node, char c) {
        leftChar[node] = c;
        rightChar[node] = c;
        prefix[node] = 1;
        suffix[node] = 1;
        max[node] = 1;
        length[node] = 1;
    }

    // In-place primitive merging to avoid generating temporary object churn
    private void merge(int parent, int a, int b) {
        leftChar[parent] = leftChar[a];
        rightChar[parent] = rightChar[b];
        length[parent] = length[a] + length[b];

        // Merge Prefix
        int aLen = length[a];
        int aPrefix = prefix[a];
        if (aPrefix == aLen && rightChar[a] == leftChar[b]) {
            prefix[parent] = aLen + prefix[b];
        } else {
            prefix[parent] = aPrefix;
        }

        // Merge Suffix
        int bLen = length[b];
        int bSuffix = suffix[b];
        if (bSuffix == bLen && rightChar[a] == leftChar[b]) {
            suffix[parent] = bLen + suffix[a];
        } else {
            suffix[parent] = bSuffix;
        }

        // Merge Maximum
        int maxVal = max[a] > max[b] ? max[a] : max[b];
        if (rightChar[a] == leftChar[b]) {
            int combo = suffix[a] + prefix[b];
            if (combo > maxVal) {
                maxVal = combo;
            }
        }
        max[parent] = maxVal;
    }
}

File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.py
class Solution:
    def longestRepeating(self, s: str, queryCharacters: str, queryIndices: list[int]) -> list[int]:
        n = len(s)
        k = len(queryIndices)
        tree_size = 4 * n

        # Flat parallel arrays to mimic Java's array cache speed locally in Python
        leftChar = [""] * tree_size
        rightChar = [""] * tree_size
        prefix = [0] * tree_size
        suffix = [0] * tree_size
        max_arr = [0] * tree_size
        length = [0] * tree_size
        leafNodeMap = [0] * n

        def build(node: int, left: int, right: int):
            length[node] = right - left + 1
            if left == right:
                leftChar[node] = s[left]
                rightChar[node] = s[left]
                prefix[node] = 1
                suffix[node] = 1
                max_arr[node] = 1
                leafNodeMap[left] = node
                return

            mid = left + (right - left) // 2
            leftChild = node << 1
            rightChild = leftChild | 1

            build(leftChild, left, mid)
            build(rightChild, mid + 1, right)

            leftChar[node] = leftChar[leftChild]
            rightChar[node] = rightChar[rightChild]

            aLen = length[leftChild]
            if prefix[leftChild] == aLen and rightChar[leftChild] == leftChar[rightChild]:
                prefix[node] = aLen + prefix[rightChild]
            else:
                prefix[node] = prefix[leftChild]

            bLen = length[rightChild]
            if suffix[rightChild] == bLen and rightChar[leftChild] == leftChar[rightChild]:
                suffix[node] = bLen + suffix[leftChild]
            else:
                suffix[node] = suffix[rightChild]

            maxVal = max_arr[leftChild] if max_arr[leftChild] > max_arr[rightChild] else max_arr[rightChild]
            if rightChar[leftChild] == leftChar[rightChild]:
                combo = suffix[leftChild] + prefix[rightChild]
                if combo > maxVal:
                    maxVal = combo
            max_arr[node] = maxVal

        build(1, 0, n - 1)

        answer = [0] * k
        for i in range(k):
            node = leafNodeMap[queryIndices[i]]
            ch = queryCharacters[i]

            leftChar[node] = ch
            rightChar[node] = ch

            node >>= 1
            while node > 0:
                leftChild = node << 1
                rightChild = leftChild | 1

                leftChar[node] = leftChar[leftChild]
                rightChar[node] = rightChar[rightChild]

                aLen = length[leftChild]
                if prefix[leftChild] == aLen and rightChar[leftChild] == leftChar[rightChild]:
                    prefix[node] = aLen + prefix[rightChild]
                else:
                    prefix[node] = prefix[leftChild]

                bLen = length[rightChild]
                if suffix[rightChild] == bLen and rightChar[leftChild] == leftChar[rightChild]:
                    suffix[node] = bLen + suffix[leftChild]
                else:
                    suffix[node] = suffix[rightChild]

                maxVal = max_arr[leftChild] if max_arr[leftChild] > max_arr[rightChild] else max_arr[rightChild]
                if rightChar[leftChild] == leftChar[rightChild]:
                    combo = suffix[leftChild] + prefix[rightChild]
                    if combo > maxVal:
                        maxVal = combo
                max_arr[node] = maxVal

                node >>= 1
                
            answer[i] = max_arr[1]

        return answer
